File: java_solutions/2609.find-the-longest-balanced-substring-of-a-binary-string.java
/*
 * @lc app=leetcode id=2609 lang=java
 *
 * [2609] Find the Longest Balanced Substring of a Binary String
 */

import java.util.regex.*;

class Solution {
    public int findTheLongestBalancedSubstring(String s) {
        return Pattern.compile("0+1+").matcher(s).results()
            .map(m -> m.group())
            .mapToInt(g -> Math.min(g.lastIndexOf('0') + 1, g.length() - g.lastIndexOf('0') - 1) * 2)
            .max()
            .orElse(0);
    }
}

File: kotlin_solutions/2609.find-the-longest-balanced-substring-of-a-binary-string.kt
/*
 * @lc app=leetcode id=2609 lang=kotlin
 *
 * [2609] Find the Longest Balanced Substring of a Binary String
 */

class Solution {fun findTheLongestBalancedSubstring(s: String) = s.split("0").filter { it.isNotEmpty() }.sortedByDescending { it.length }.let { ones -> s.split("1").filter { it.isNotEmpty() }.sortedByDescending { it.length }.flatMap { zero -> ones.map { one -> if (s.contains(zero + one)) minOf(zero.length, one.length) * 2 else 0 } } }.maxOrNull() ?: 0}

